vms/platformvm/dao_add_vote_tx.go

```diff
@@ -0,0 +1,220 @@
+func (tx *UnsignedDaoAddVoteTx) InitCtx(ctx *snow.Context) {
+	tx.BaseTx.InitCtx(ctx)
+}
+
+// SyntacticVerify returns nil if [tx] is valid
+func (tx *UnsignedDaoAddVoteTx) SyntacticVerify(ctx *snow.Context) error {
+	switch {
+	case tx == nil:
+		return errNilTx
+	case tx.syntacticallyVerified: // already passed syntactic verification
+		return nil
+	}
+
+	if err := tx.BaseTx.SyntacticVerify(ctx); err != nil {
+		return fmt.Errorf("failed to verify BaseTx: %w", err)
+	}
+
+	// cache that this is valid
+	tx.syntacticallyVerified = true
+	return nil
+}
+
+// Attempts to verify this transaction with the provided state.
+func (tx *UnsignedDaoAddVoteTx) SemanticVerify(vm *VM, parentState MutableState, stx *Tx) error {
+	_, _, err := tx.Execute(vm, parentState, stx)
+	return err
+}
+
+// Execute this transaction.
+func (tx *UnsignedDaoAddVoteTx) Execute(
+	vm *VM,
+	parentState MutableState,
+	stx *Tx,
+) (
+	VersionedState,
+	VersionedState,
+	error,
+) {
+	// Verify the tx is well-formed
+	if err := tx.SyntacticVerify(vm.ctx); err != nil {
+		return nil, nil, err
+	}
+
+	daoProposals := parentState.DaoProposalChainState()
+	currentStakers := parentState.CurrentStakerChainState()
+
+	var totalTransferable uint64
+
+	for _, in := range tx.Ins {
+		if in.AssetID() == vm.ctx.AVAXAssetID { // only allow to use avax to bond
+			totalTransferable += in.In.Amount()
+		}
+	}
+
+	if totalTransferable < vm.DaoConfig.VoteTxFee {
+		return nil, nil, fmt.Errorf("provided ins do not cover for vote fee")
+	}
+
+	if vm.bootstrapped.GetValue() {
+		currentTimestamp := parentState.GetTimestamp()
+		// Ensure the proposed validator starts after the current time
+		startTime := tx.StartTime()
+		if !currentTimestamp.Before(startTime) {
+			return nil, nil, fmt.Errorf(
+				"daoVote start time (%s) at or before current timestamp (%s)",
+				startTime,
+				currentTimestamp,
+			)
+		}
+
+		validator, err := currentStakers.GetValidator(tx.NodeID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// verify that the proposal is active and that we have not voted so far
+		daoProposal, err := daoProposals.GetActiveProposal(tx.ProposalID)
+		switch {
+		case err != nil:
+			return nil, nil, err
+		case !currentTimestamp.Before(daoProposal.DaoProposalTx().EndTime()):
+			return nil, nil, fmt.Errorf("proposal: %s is not active anymore", tx.ProposalID.String())
+		case len(daoProposal.Votes()) >= int(daoProposal.DaoProposalTx().ProposalConfiguration.Thresh):
+			return nil, nil, fmt.Errorf("proposal: %s already accepted", tx.ProposalID.String())
+		case daoProposal.Voted(tx.NodeID):
+			return nil, nil, fmt.Errorf("node %s has already voted on proposal: %s", tx.NodeID.String(), tx.ProposalID.String())
+		case daoProposal.DaoProposalTx().StartTime().After(validator.AddValidatorTx().StartTime()):
+			return nil, nil, fmt.Errorf("node %s only became a validator after the proposal started and is thus not eligible to participate", tx.NodeID.String())
```

Comment:
That should rather be:
`daoProposal.DaoProposalTx().StartTime().**Before**(validator.AddValidatorTx().StartTime()):`